Sources/Mentions.php

```diff
@@ -234,6 +301,137 @@ protected static function getPossibleMentions($body)
 	}
+
+	/**
+	 * Like getPossibleMentions(), but for `[member=1]name[/member]` format.
+	 *
+	 * @static
+	 * @access public
+	 * @param string $body The text to look for mentions in.
+	 * @param array $members An array of arrays containing info about members (each should have 'id' and 'member').
+	 * @return array An array of arrays containing info about members that are in fact mentioned in the body.
+	 */
+	public static function getExistingMentions($body)
+	{
+		// Don't include mentions inside quotations.
+		$body = preg_replace('~\[quote[^\]]*\](?' . '>(?' . '>[^\[]|\[(?!/?quote[^\]]*\]))|(?0))*\[/quote\]~', '', $body);
+
+		$existing_mentions = array();
+
+		preg_match_all('~\[member=([0-9]+)\]([^\[]*)\[/member\]~', $body, $matches, PREG_SET_ORDER);
+
+		foreach ($matches as $match_set)
+			$existing_mentions[$match_set[1]] = trim($match_set[2]);
+
+		return $existing_mentions;
+	}
+
+	/**
+	 * Verifies that members really are mentioned in the text.
+	 *
+	 * This function assumes the incoming text has already been processed by
+	 * the Mentions::getBody() function.
+	 *
+	 * @static
+	 * @access public
+	 * @param string $body The text to look for mentions in.
+	 * @param array $members An array of arrays containing info about members (each should have 'id' and 'member').
+	 * @return array An array of arrays containing info about members that are in fact mentioned in the body.
+	 */
+	public static function verifyMentionedMembers($body, array $members)
+	{
+		if (empty($body))
+			return array();
+
+		// Don't include mentions inside quotations.
+		$body = preg_replace('~\[quote[^\]]*\](?' . '>(?' . '>[^\[]|\[(?!/?quote[^\]]*\]))|(?0))*\[/quote\]~', '', $body);
+
+		foreach ($members as $member)
+		{
+			if (strpos($body, '[member=' . $member['id'] . ']' . $member['real_name'] . '[/member]') === false)
+				unset($members[$member['id']]);
+		}
+
+		return $members;
+	}
+
+	/**
+	 * Retrieves info about the authors of posts quoted in a block of text.
+	 *
+	 * @static
+	 * @access public
+	 * @param string $body A block of text, such as the body of a post.
+	 * @param int $poster_id The member ID of the author of the text.
+	 * @return array Info about any members who were quoted.
+	 */
+	public static function getQuotedMembers($body, $poster_id)
+	{
+		global $smcFunc;
+
+		if (empty($body))
+			return array();
+
+		$blocks = preg_split('/(\[quote.*?\]|\[\/quote\])/i', $body, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
+
+		$quote_level = 0;
+		$message = '';
+
+		foreach ($blocks as $block)
+		{
+			if (preg_match('/\[quote(.*)?\]/i', $block, $matches))
+			{
+				if ($quote_level == 0)
+					$message .= '[quote' . $matches[1] . ']';
+				$quote_level++;
+			}
+			elseif (preg_match('/\[\/quote\]/i', $block))
+			{
+				if ($quote_level <= 1)
+					$message .= '[/quote]';
+				if ($quote_level >= 1)
+				{
+					$quote_level--;
+					$message .= "\n";
+				}
+			}
+			elseif ($quote_level <= 1)
+				$message .= $block;
+		}
```

Comment:
What is the purpose of this code ? I can't wrap my head around it.

Reply:
This is just copied straight over from its previous location in CreatePost-Notify.php. It appears to be walking through all matches from the regex above in order to make sure nested quotes do not mess it up. The whole thing could probably be replaced with a properly recursive regex like added elsewhere in this PR, but I didn't want to mess with it when it currently isn't broken.